### PR TITLE
Move/Fix FakeTensor logic for detecting multiple fake modes

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97186

This was leftover for when we had more logic in the FakeTensor and not FakeTensorMode, and wasn't firing correctly. It also makes more sense for it to be in the other validation function.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by bdhirsh — APPROVED** [review]
this is great, would have saved me some debugging pain a while ago

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (BowenBao) [inline torch/onnx/_internal/fx/passes/decomp.py:51]: cc @wschin

File: torch/fx/experimental/proxy_tensor.py
Comment (eellison) [inline torch/fx/experimental/proxy_tensor.py:660]: cc @Chillee @bdhirsh @voznesenskym @ezyang 

i made these api changes. feel free to bikeshed.

File: torch/onnx/_internal/fx/passes/shape_inference.py
Comment (BowenBao) [inline torch/onnx/_internal/fx/passes/shape_inference.py]: nit
```suggestion
    fake_tensor_mode = torch._dynamo.utils.fake_mode_from_tensors(args) or fake_tensor.FakeTensorMode()
```

Comment (wschin) [reply torch/onnx/_internal/fx/passes/shape_inference.py]: I remember we want to use the `init_env` in FakeTensorMode to override some shapes. If we can't create FakeTensorMode here, then we need to modify FakeTensorMode's API to add a flag. @titaiwangms, for heads up.

File: torch/fx/experimental/proxy_tensor.py
Comment (wschin) [inline torch/fx/experimental/proxy_tensor.py:661]: `fake` and `existing` looks duplicated. When caller specifying `existing`, I assume the code will create FakeTensorMode if it doesn't exist. Since there is no document, I don't quite understand the design concepts and expected behaviors of those options.

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (eellison) [inline torch/onnx/_internal/fx/passes/decomp.py:40]: @bdhirsh @ezyang this is "wrapping" of a callable so that users can trace with an existing fake mode. IMO, it is a lot to expect a user to write, especially considering the number of times we've seen a user attempt to do something like this themselves.... however i'm really just trying to unblock the fake tensor error so I will do whatever to land this

**Comment by eellison** [comment]
cc @wschin @BowenBao I'm trying to make a change that will error when multiple FakeTensorModes are active, since this is not supported and leads to confusing errors currently. On the pr there were onnx failures because it would call `make_fx` with FakeTensor inputs, so there would exist both the mode from the inputs and the newly constructed fake tensor mode. I have implemented a wrapper so that there would not be multiple fake modes around. Now i'm getting other onnx errors I dont know how to interpret/fix. Would you please take a look ?

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (BowenBao) [inline torch/onnx/_internal/fx/passes/decomp.py:50]: @eellison onnx error from this line. `mode` is a str it appears?

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (BowenBao) [inline torch/onnx/_internal/fx/passes/decomp.py:40]: Related to onnx error comment. Type annotating the arguments helps bring a lot clarity.

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/decomp.py:50]: guess should be `fake_mode`?

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (ezyang) [reply torch/onnx/_internal/fx/passes/decomp.py:40]: I agree we need to work this out a better way to do this. Maybe we can discuss this tomorrow at composability social. My current thinking is that if we can enforce there is one fake mode, it makes sense to have an easy context manager `with_fake` and `no_fake`.

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (eellison) [reply torch/onnx/_internal/fx/passes/decomp.py:50]: Thanks!

Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/decomp.py]: @wschin for `export_without_parameters_and_buffers`

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (wschin) [inline torch/onnx/_internal/fx/passes/decomp.py:110]: `real`? I guess we always want to run decomposition with `fake` or `symbolic`. Maybe you can add another flag `reuse_existing_fake_or_create_fake`.?

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (wschin) [reply torch/onnx/_internal/fx/passes/decomp.py]: I don't have clear idea. I need to see how the `input_` passed into that `assert`. It might be related to argument-wrapping code in exporter. If I get sometime this week, I will try debug.

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (eellison) [reply torch/onnx/_internal/fx/passes/decomp.py:110]: That was the flag I tried to add which people didn't want me to lol. There may be changes in the future to the api after this pr.

Comment (wschin) [reply torch/onnx/_internal/fx/passes/decomp.py:110]: Humm, I like that flag. If we change the value to `real`, exporter will be way slower.

Comment (eellison) [reply torch/onnx/_internal/fx/passes/decomp.py:110]: we're not actually doing real, we're still wrapping the inputs to fake. We're avoiding calling `real` so we dont do a new FakeTensorMode on top of the mode from the inputs.

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (wschin) [reply torch/onnx/_internal/fx/passes/decomp.py]: I am building this branch. Let's see if I can reproduce it tomorrow.

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (wschin) [reply torch/onnx/_internal/fx/passes/decomp.py:110]: What if the input must be a FakeTensor?

In the error above, the whole model is initialized under FakeTensorMode, so all its parameters/buffers are FakeTensor's already. I guess that's why you see that assert. I am not sure how to deal with it without running the code. Building PyTorch on my side will take a while.

Comment (eellison) [reply torch/onnx/_internal/fx/passes/decomp.py:110]: need any pointers ?

**Comment by eellison** [comment]
cc @bowenbao @wschin can you please a take a look ?

File: torch/onnx/_internal/fx/passes/decomp.py
Comment (BowenBao) [inline torch/onnx/_internal/fx/passes/decomp.py:44]:
```suggestion
    if fake_mode is None:
```
I think this resolves the onnx errors in CI. I'm not sure about the one you posted though, building locally to have a look.

**Comment by BowenBao** [comment]
@eellison with new changes I'm getting all test cases failing under the test file, even for ones that were not creating their own fake tensor mode. I'm not sure how to deal with those.
```
---  TestFxToOnnxWithOnnxRuntime.test_simple_function ---
...
  File "/home/bowbao/pytorch/torch/fx/experimental/proxy_tensor.py", line 99, in get_proxy_slot
    raise RuntimeError(f"{obj} is not tracked with proxy for {tracer}")
RuntimeError: s0 is not tracked with proxy for <torch.fx.experimental.proxy_tensor.PythonKeyTracer object at 0x7f5423302190>

While executing %add : [#users=2] = call_function[target=operator.add](args = (%x, 1.0), kwargs = {})
Original traceback:
  File "/home/bowbao/pytorch/test/onnx/test_fx_to_onnx_with_onnxruntime.py", line 110, in func
    y = x + 1.0
```

For the other one that creates its own fake tensor mode outside, I guess the solution needs to be extended to look for fake tensor mode from model weights as well.

**Comment by BowenBao** [comment]
@eellison so the problem is with how "user code" stash up the contexts outside. The current code below doesn't work because it does not accurately resemble what happens inside `make_fx`.
```python
    def fn_wrapped():
        with context():
            with fake_mode:
                return fn(*args)
```
Instead, it should be something like
```python
        fake_args, shape_env, fake_mode, context = wrapped_fn(
            graph_with_interpreter, fx_mode, allow_non_fake_inputs=True, args=args
        )

        with fake_mode, context():
            decomposed_module = proxy_tensor.make_fx(
                graph_with_interpreter,
                decomposition_table=self.decomposition_table,
                tracing_mode="real",
                _allow_non_fake_inputs=True,
            )(*fake_args)
        if fx_mode == "symbolic":
            decomposed_module.shape_env = shape_env
```

Frankly the issue had nothing to do with onnx. With this change I'm passing the onnx tests. tbh this looks very messy, I'm looking forward to the general fake_mode follow up. cc @ezyang

**Comment by ezyang** [comment]
At composability sync, the tentative agreement was that we would make make_fx, by default, automatically detect if you pass in fake inputs or have fake mode and use that directly; or if there was no fake mode automatically fakeify.

**Comment by ezyang** [comment]
What's holding up landing this?

**Comment by BowenBao** [comment]
@eellison could you try update the PR with my suggestions above? That should unblock the onnx CI failures.

File: torch/fx/experimental/proxy_tensor.py
Comment (eellison) [inline torch/fx/experimental/proxy_tensor.py:701]: how do these changes look @ezyang ?

Comment (ezyang) [reply torch/fx/experimental/proxy_tensor.py:701]: Yep, looks great!

**Comment by BowenBao** [comment]
@eellison do you have a link to the failing onnx ci log?

**Comment by BowenBao** [comment]
Error is from `torch/onnx/_internal/fx/passes/shape_inference.py` where a new `fake_tensor_mode` was created and used before calling `FakeTensorPropGetStaticShapes(module, fake_tensor_mode).propagate(*args)`. Looks like now the proper way should be  try grabbing existing mode by `fake_tensor_mode = torch._dynamo.utils.detect_fake_mode(args)` before creating the new mode. Is that correct @eellison ?

cc @titaiwangms

**Comment by eellison** [comment]
@BowenBao yes that's correct

**Comment by BowenBao** [comment]
Rebase with master and apply this diff should unblock.

```diff
diff --git a/torch/onnx/_internal/fx/passes/shape_inference.py b/torch/onnx/_internal/fx/passes/shape_inference.py
index a708f354509..14a41a8f277 100644
--- a/torch/onnx/_internal/fx/passes/shape_inference.py
+++ b/torch/onnx/_internal/fx/passes/shape_inference.py
@@ -4,6 +4,7 @@ import itertools
 from typing import Optional
 
 import torch
+import torch._dynamo
 import torch.fx
 from torch._subclasses import fake_tensor
 from torch.fx.experimental import proxy_tensor
@@ -33,7 +34,9 @@ class ShapeInferenceWithFakeTensor(_pass.Transform):
         # 2. run FakeTensorProp
         # If (1) and (2) are done with difference FakeTensorMode's, undefined behavior may
         # happen.
-        fake_tensor_mode = fake_tensor.FakeTensorMode()
+        fake_tensor_mode = (
+            torch._dynamo.utils.detect_fake_mode(args) or fake_tensor.FakeTensorMode()
+        )
 
         def to_fake_tensor(x):
             if isinstance(x, torch.Tensor) and not isinstance(
```

**Pull request closed without merging**
